sandbox/apparmor/apparmor.go

```diff
@@ -56,6 +57,221 @@ func ValidateNoAppArmorRegexp(s string) error {
+// excludePatterns will match with the specified prefix and suffix rules. For
+// example to allow reading any file except those matching /usr/*/foo, you would
+// call this function with "/usr/*/foo" as the first argument, "" as the prefix
+// and " r," as the suffix (the suffix being the main part of the read rule) and
+// this function would return the following multi-line string with the relevant
+// rules:
+//
+// /[^u]** r,
+// /u[^s]** r,
+// /us[^r]** r,
+// /usr[^/]** r,
+// /usr/*/[^f]** r,
+// /usr/*/f[^o]** r,
+// /usr/*/fo[^o]** r,
+//
+// This function only treats '*' specially in the string and does not handle any
+// other alternations etc. that AARE may more generally support, and all
+// patterns provided must be absolute filepaths that are at least 2 runes long.
+//
+// This function also works with multiple exclude patterns such as specifying to
+// exclude "/usr/lib/snapd" and "/var/lib/snapd" with suffix " r," would yield:
+//
+// /[^uv]** r,
+// /{u[^s],v[^a]}** r,
+// /{us[^r],va[^r]}** r,
+// /{usr[^/],var[^/]}** r,
+// /{usr/[^l],var/[^l]}** r,
+// /{usr/l[^i],var/l[^i]}** r,
+// /{usr/li[^b],var/li[^b]}** r,
+// /{usr/lib[^/],var/lib[^/]}** r,
+// /{usr/lib/[^s],var/lib/[^s]}** r,
+// /{usr/lib/s[^n],var/lib/s[^n]}** r,
+// /{usr/lib/sn[^a],var/lib/sn[^a]}** r,
+// /{usr/lib/sna[^p],var/lib/sna[^p]}** r,
+// /{usr/lib/snap[^d],var/lib/snap[^d]}** r,
+//
+// Note that with the previous rules, /usr/lib/snapdaemon would not match any rule
+//
+// This function has the following limitations:
+// The first character after a subpattern common to two or more excludePatterns cannot
+// be '*' on any of the excludePatterns that share the common prefix.
+// Eg. ["/snap/core/**", "/snap/*/core/**"] where '/snap/' would be the subpattern common
+// to both excludePattern, and '*' would be the first character after the common subpattern
+// in the second excludePattern.
+// This is because there are no apparmor rules that can fulfill both requirements.
+// For /snap/[^c]** -> It will also match /snap/a/core/** which should be excluded by
+// the second pattern.
+// For /snap/[^c*]** -> It will also exclude access to /snap/a/a that should be allowed
+// as it is not explicitly excluded by any pattern
+//
+// When the '*' is used to exclude suffixes, like in ["/*.bin"], rules should be generated
+// in a reverse way:
+//
+// /*[^n]{,/**} rw,
+// /*[^i]n{,/**} rw,
+// /*[^b]in{,/**} rw,
+// /*[^.]bin{,/**} rw,
+//
+// While generating those rules is technically possible, it will make the logic way more
+// complex, thus the function would just return an error if a pattern of this kind is found.
+// This functionality can be added in a subsequent PR if needed in the future
+func GenerateAAREExclusionPatterns(excludePatterns []string, opts *AAREExclusionPatternsOptions) (string, error) {
+	if len(excludePatterns) == 0 {
+		return "", fmt.Errorf("no patterns provided")
```

Comment:
This should be `errors.New`

Reply:
Done in commit 25f75be5889fc9b06f51856734f4fb712dfe42fa